### DISCUSSION TITLE
What's the difference between Router.push and (useRouter().push)

### DISCUSSION THREAD

**Discussion by wildskyf** [vercel/next.js#18522] [Help]
This
```javascript
import Router, from 'next/router'

const MyComponent = () => {
  // whatever
  const doSomething = () => Router.push('A LINK')

  return (
      // whatever
  )
}
```

and this
```javascript
import { useRouter } from 'next/router'

const MyComponent = () => {
  const router = useRouter()
  // whatever
  const doSomething = () => router.push('A LINK')

  return (
      // whatever
  )
}
```

both works, but is there any difference between them?

**Comment by MathiasKandelborg** [comment]
There isn't, it's just a matter of different approaches.

You can also[ get access to `router` via a HoC](https://nextjs.org/docs/api-reference/next/router#withrouter)


I don't think the direct Router import is a suggested approach. There's no real reason to import that instead of using either a hook or HoC.

**Reply by kartikag01** [reply]
@timneutkens
if there isn't any difference, then directly import `Router` is much easier in both class component and functional components, why recommended approach is `useRouter().push`  ?

**Comment by omariosouto** [comment]
@wildskyf I personally prefer to use the useRouter approach and refactor the component from class to function, considering that some times I face situations where I need to use `router.isReady`

> isReady: boolean - Whether the router fields are updated client-side and ready for use. Should only be used inside of useEffect methods and not for conditionally rendering on the server.

https://nextjs.org/docs/api-reference/next/router

**Reply by luiznasciment0** [reply]
100 times thank you ❤️ I thought there was technical differences but it seems to be more personal than technical

**Comment by willianjusten** [comment]
As the others said, it's just a matter of different approaches, but you need to keep in mind:

> You cannot use `useRouter` hook inside Class Components, since hooks works `only` in functional components

So that's the main reason to have both.

---

If you look the code itself, you'll see that the `useRouter` is just a `RouterContext` to the Class, so in the end, both will have same functionalities, the object is the same.

`useRouter`: https://github.com/vercel/next.js/blob/canary/packages/next/client/router.ts#L127-L129

`Router`: https://github.com/vercel/next.js/blob/canary/packages/next/next-server/lib/router/router.ts#L505

**Reply by luiznasciment0** [reply]
🙏🏼🙏🏼❤️ thanks

**Comment by devknoll** [comment]
Note: for the best compatibility with e.g. concurrency in `react@>=18`, you should generally:

1. Only use the global `Router` object outside of the React lifecycles (i.e. inside `useEffect` or event handlers)

2. Use `useRouter` in function components, or `withRouter` for Class components

In general, the difference is that the global `Router` object always points to the latest global state at _any given time_, whereas `useRouter`/`withRouter` refer to _the route being rendered_.

An easier way to think about it might be that `useRouter`/`withRouter` are actually `route`s that happen to also include router functions (like `push`), where the global `Router` object is a bundle of router functions (like `push`) that happens to also include the latest state for the current `route`.

**Reply by devknoll** [reply]
This distinction is particularly important for concurrent features. Consider what happens if a user clicks to load a new page, but that page suspends because it needs to fetch more data, and while waiting the user clicks on a button that should `alert()` the current pathname.

* The first new page will be rendered concurrently. The page will see a different `useRouter` result that corresponds to the new page so components can correctly read e.g. the url params to fetch any new data or whatever 
* The new click event on the old page needs to see the router that’s currently on screen, so that it can display the correct pathname

**Comment by aboveyunhai** [comment]
@devknoll  based on your clarification it's the following considered as a fair use? I usually faced the following scenario when try to shallow push the new url;
The issue case: 
it will have the following warning as expect
`React Hook useEffect has a missing dependency: 'router'. Either include it or remove the dependency array.eslintreact-hooks/exhaustive-deps`
```javascript
const router = useRouter();
useEffect(() => {
  // console.log(Router.query.depRaram);
  if (depParams) { 
     router.push( `?exampleParam=${depParam}`, undefined, { shallow: true })
   }
}, [depParam])
```
I don't want to add router as dependency here, `router.push` will cause infinite loop becuase it change `router`.
 something like `[depParam,  deepObjCheck(router)]` also felt not right for me.

This works as expect without warning since we don't use `useRouter`
```javascript
useEffect(() => {
  // console.log(Router.query.depRaram);
  if (depParams) { 
     Router.push( `?exampleParam=${depParam}`, undefined, { shallow: true })
   }
}, [depParam])
```
Also, I wonder will I any trouble if I want to use `Router.query.depRaram` directly inside `useEffect`.

**Reply by devknoll** [reply]
Sorry for the delay, for some reason I didn't see this until now. Yes, using the global `Router` to set a new state is fine. However, in this case you should likely resolve this issue by just checking if `exampleParam` is already set to `depParam` to avoid the infinite loop, instead of worrying about `router`, for the second reason:

> Also, I wonder will I any trouble if I want to use `Router.query.depRaram` directly inside `useEffect`.

I recommend always reading values from `useRouter` when inside of a component as a best practice. Since the global `Router` is... global, it could be the case that some React hooks could be called before it is updated, and some could be called after it's updated but before your component has re-rendered (such as `useLayoutEffect` and `useInsertionEffect`). Therefore it's best to just use `useRouter` to ensure you always get the value expected when the component was rendered.

**Comment by hoatep** [comment]
I don't know how to operate by router and useRouter() but when i use router  to get the error.

![image](https://user-images.githubusercontent.com/86640624/153121733-c0fc26ce-54d6-4958-aef4-e27356a1d3d0.png)

In _app.ts file, I have function split case ssr and csr

<img width="419" alt="image" src="https://user-images.githubusercontent.com/86640624/153122044-b8d4c1fa-bdf4-40a0-b319-233c67076a59.png">

Temporary solution: replace router by useRouter() and it's work
Anyone with an explanation for this problem, welcome

**Reply by mlarcher** [reply]
IIRC that happens when you use your IDE to add the import for you, and it ends up being `import { router } from 'next/client'` instead of `import { router } from 'next/router'`. You should double check on that, though

**Comment by dwilt** [comment]
> There isn't, it's just a matter of different approaches.

Our team has found the hard way that there is absolutely a difference. When you use the `useRouter` hook, you are buying into re-renders _whenever the URL changes_ because the internal state of the `useRouter` hook changes.

So in the case of using just a `push` method, we _only_ use the router singleton (`import Router from 'next/router'`) because this import _is stable_. In fact, we have internal dev tasks to remove _all_ instances of `useRouter` to avoid any additional re-renders since most of the time, we access the URL _after_ the client has loaded and we don't need the components being bound to the URL changes.

**Reply by icyJoseph** [reply]
This had always been my answer, I didn't really know that people thought these were technically the same. 

While I do see your point, re-rendering is not inherently an issue. There's also legit use cases for `useRouter` and `Router`, and it is not as crystal clear as, "remove all" instances. There's also techniques to isolate hooks from others via Render Props.

I have pages that use `getStaticPaths` with fallback set `true`, and to switch between fallback UI and formal UI, I need to check `useRouter().isFallback`, Router in this case just crashes out, even if you check for readiness.

On the contrary, if I want to use router logic outside React functional components, like invoking `NProgress`, I can just do that with Router, anywhere, at the top of `_app` for instance, or perhaps I want to call an analytics tracker.

```tsx
Router.events.on("routeChangeStart", () => NProgress.start());
Router.events.on("routeChangeComplete", () => NProgress.done());
Router.events.on("routeChangeError", () => NProgress.done());

function App() {
useEffect(() => {
    const handler = (url: string) => pageview(url);

    Router.events.on("routeChangeComplete", handler);

    return () => Router.events.off("routeChangeComplete", handler);
  }, []);
```

Reaching out for `push` is also something to keep under the control. If it is within control of my `api/` routes I prefer to redirect form submissions from there.

If you need to run side-effect logic that takes the user to another page, why are renders a problem? The components should un-mount on page change, unless you are using the router in `_app`? Either directly or through shared layout components.

```tsx
// _app.tsx
  const router = useRouter();

  useEffect(() => {
    // executes forever
    router.push("/");
  }, [router]);
```

In such cases, you probably do want to inform your components about URL changes, but there's other conditions that should limit how the router triggers? I am thinking for example, a **Header** that checks if there's a user present in state, and if not, redirects to login, but that **redirect** effect should only execute, if the current path, is not the **login** path, right? 

Of course there's the case where a `useEffect` depends on router, and some other data, but when the URL changes, you run the effect again, though you probably wanted to run it only if data changed, then it is best to use `Router`.

```tsx
  // inside a component that's present in all routes
  const data = useMemo(() => ({}), []);

  useEffect(() => {
    console.log("D"); // this logs every time the router changes, which is probably unwanted, we care for data only!   
    router.push("/something") // if we don't add an extra check for router.pathname === 'something'
   // this turns into a run away effect :( 
  }, [router, data]);

  useEffect(() => {
    console.log("D");// only logs when data changes
    Router.push('/something'); // stops right after this
  }, [data]);
```

I am making these points because it seems like you've taken a very stark decision, as in "remove all", and given that **Discussions** are often used by people who are learning or trying to solve an issue, your answer might confuse people, rarely there's such a silver bullet.

I am not saying you are wrong, or that `router.push` doesn't have legit use cases, or that things are problem free, but our job as developers, is to assess a situation and respond as best as we can with the tools the framework gives us.

I am also curious to know the "hard way" you learned this, I like to document these things and keep them on my archive.

**Reply by dwilt** [reply]
Let me try and add some clarity which will hopefully resolve any confusion devs might've taken from my previous answer.

![image](https://user-images.githubusercontent.com/1065539/159764955-b6587458-7eb0-4be9-adeb-473d1bf853de.png)

Here's a screenshot of our application. Almost all of the components that you see in the screen (links/items in the left nav, the header and in the content area) have a reference to the router inside of them. Most of them access the `push` method. The problem is that when we used the native `<Link />` element or the `push` from `useRouter`, _elements that persisted between pages re-rendered_ (even if query params were just adjusted).

We frequently test our application in 6x slowdown to simulate slower computers. Since we have so many components on the screen, there were actual measurable (and noticeable as a user) differences from the reconciliation process that React had to run through over those components. Hence why we reached for the singleton approach. Our application has a lot of components with complex rendering (drag and drop along with large images and videos) so trying to reduce re-renders (as well as improve the speed of them) is a high priority for us.

**Reply by aboveyunhai** [reply]
@dwilt  though performance is not main issue I faced but I think we hit the same limitation of `useRouter()` in the core. That's why I asked the same question of using `Router` previously.

To sum up:

`Component` **update/action** will conditionally change `Router`, but anything happens in `Router` should never fire component action, not even call the `useEffect()`. Checking `url` differences cannot solve many issues when `rewrite` kicks in, and it unnecessarily calles `useEffect()` and complicates the codebase.

`Router` is **Unidirectional**, `useRouter()` is a **Bidirectional** in contrast. 

This also makes sure next `rewrite` wouldn't cause any weird behavior to the `useEffect()` in some rare cases, especially when the components are highly associated with router shallow update.

**Comment by remjx** [comment]
In my application `router.push` was being called twice in the same render cycle and neither had any effect. I don't know why, but after I removed one of them, `router.push` works as expected 🤷‍♂️

(Sorry if this comment is out of context or I am duplicating a comment above, but I figured this out after reading this thread.)

**Comment by dovranJorayev** [comment]
And what is the problem? Don't put router as immutable dependency as it is immutable object each time route change. For example use custom hook with router ref.
```
export const useRouterRef = () => { 
    const router  = useRouter();
    const routerRef = React.useRef(router);
    routerRef.current = router;
    return routerRef;
}

const Component = () => {
   const routerRef = useRouterRef();
   const externalData = useExternalDataQuery();
   
   useEffect(() => {
     routerRef.current.push(`/some-url/${externalData.id}`);
     }, [externalData,  routerRef]);  // Add here only necessary things. P.S. routerRef is mutable and don't cause re-renders
     
     return (...);
}
```
Also as option you can disable eslint with plain router and don't use `router` as dependency

**Reply by dwilt** [reply]
@dovranJorayev I believe this will still cause you a re-render every time the URL changes because `useRouter` will re-render and give you a new `router` value. Then any components/hooks that utilize this hook, will re-render.

**Reply by dovranJorayev** [reply]
Re-renders don't re-run useEffect here unless `externalData` change. Cause useRouterRef will return mutable object reference it will not re-run useEffect.  Also `externalData` here is an object and could be bad example. But if we will change this to this
```
 useEffect(() => {
     routerRef.current.push(`/some-url/${externalData.id}`);
     }, [externalData.id,  routerRef]);
 ```
 Effect never run unless externalData.id changes as supposed to be.

**Reply by dwilt** [reply]
Correct, but `Component` will re-render.

**Reply by dovranJorayev** [reply]
You right, but I am not afraid of re-renders, i am afraid unexpected behavior like useless effect calls, but with react 18 (dev mode) it especially the problem when you expect it run once. As [react team recommends](https://react.dev/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development) we need write logic that way that not depend of effects calls count (include clean up or other tradeoffs).

I believe in general writing such logic that depend on View engine's (react) render cycle is anti pattern that need to avoid. Usually it can be placed right after some event that caused the update of `externalData.id` and do not do it in useEffect
